### PR TITLE
Rework to match UW Endpoint 

### LINKED ISSUE #15: Modify Group LMS User Sync Settings Form
Acceptance Criteria:

- [ ] Secret Key Field: Change to "Public Key"
- [ ] Add New Field: "Private Key" (similar to Public Key)
- [ ] API Version: Change to "Endpoint URL"

This form will hold the following data:

- The public key value (from the Key module)
- The private key value (from the Key module)
- The base URL to the endpoint (e.g. https://www.someurl.com/api/1.0/something/)

The base URL will be concated to values from private/public key and other values. Refer to ticket for Endpoint URL ().

### LINKED ISSUE #16: Use API endpoint for UW Middleware
Endpoint URL for this client will be:

`https://www.serverurl.com/api/1.0/users/roles/unit_id/get/key/{{PUBLIC_KEY}}/hash/{{PRIVATE_KEY_HASHED}}/format/json/unit_id/{{OrgUnitId}}/`

PUBLIC_KEY and PRIVATE_KEY will come from key module and lms sync module settings.

Hashed version of private key to be used in endpoint like so: `$hashed_key = hash('sha256', $private_key . 'json');`

So the request might look like this:

```
// $this->endpoint_url = 'https://www.serverurl.com/api/1.0/users/roles/unit_id/get/key/';
// $pub_key = $this->repository->getKey($this->endpoint_publickey)->getKeyValue();
// $priv_key = $this->repository->getKey($this->endpoint_privatekey)->getKeyValue();
// $hashed_key = hash('sha256', $private_key . 'json');

$request = $client->get($this->endpoint_url . $priv_key . '/hash/' . $hashed_key . '/format/json/unit_id/' . $group_id;

// outputs something like this: https://www.serverurl.com/api/1.0/users/roles/unit_id/get/key/66a294130e3bfbd744f7c275b2899e429acc78e6a0406a2ce696fc9f37ecebd7/hash/a1913ff83f9c32b0f4b3f1dc36f30977c16788cacad9a643acc72e77840b533f/format/json/unit_id/1017765/
```

### LINKED ISSUE #17: Adjust to UW returned data
The UW endpoint will return data in the following format (note that it's an associative array that inheritly groups by role.id):

```
[
    {
        "1234": {
            "id": 1234,
            "user_id": "987",
            "username": "ydahi",
            "display_name": "Yasin Dahi",
            "first_name": "Yasin",
            "last_name": "Dahi",
            "role": {
                "id": 110
            }
        }
    },
    {
        "185": {
            "id": 185,
            "user_id": "textvalue",
            "username": "jsmith",
            "display_name": "Jim Smith",
            "first_name": "Jim",
            "last_name": "Smith",
            "role": {
                "id": 109
            }
        },
        "111": {
            "id": 111,
            "user_id": "",
            "username": "somename",
            "display_name": "Some Name",
            "first_name": "Some",
            "last_name": "Name",
            "role": {
                "id": 109
            }
        }
    },
    {
        "123": {
            "id": 123,
            "user_id": "123",
            "username": "jstudent",
            "display_name": "Jessica Student",
            "first_name": "Jessica ",
            "last_name": "Student",
            "role": {
                "id": 106
            }
        }
    },
....
```

Important  Fields:

1. username: use this value to match against drupal usernames to see if account exists.
2. role.id: use this value to add the correct Drupal Role and the correct Group Role.

Role.ID mapping is provided in a separate issue.

### LINKED ISSUE #18: Map Role.ID to Drupal Role and Group Role
The UW Endpoint uses the following values for Role.ID:

```
ROLE_ID_ADMINISTRATOR                = 102;
ROLE_ID_COURSE_MANAGER               = 110;
ROLE_ID_CEL_COURSE_EDITOR            = 111;
ROLE_ID_FUTURE_STUDENT               = 124;
ROLE_ID_STUDENT                      = 107;
ROLE_ID_STAFF                        = 113;
ROLE_ID_INSTRUCTOR                   = 106;
ROLE_ID_TA_LEVEL_4                   = 112;
ROLE_ID_TA_LEVEL_3                   = 115;
ROLE_ID_TA_LEVEL_2                   = 117;
ROLE_ID_TA_LEVEL_15                  = 126;
ROLE_ID_TA_LEVEL_1                   = 114;
```

Map the following:

ROLE_ID_ADMINISTRATOR (102): 
Drupal Role: Staff
Group Role: View

ROLE_ID_COURSE_MANAGER (110):
Drupal Role: Staff
Group Role: Creator

ROLE_ID_CEL_COURSE_EDITOR (111):
Drupal Role: CEL_Staff
Group Role: Creator

ROLE_ID_FUTURE_STUDENT  (124):
Drupal Role: Student
Group Role: View

ROLE_ID_STUDENT (107):
Drupal Role: Student
Group Role: Student

ROLE_ID_STAFF (113)
Drupal Role: Staff
Group Role: Content Editor

ROLE_ID_STAFF (113)
Drupal Role: Staff
Group Role: Content Editor

ROLE_ID_INSTRUCTOR (106)
Drupal Role: Faculty
Group Role: View

ROLE_ID_TA_LEVEL_4 (112),
ROLE_ID_TA_LEVEL_3 (115),
ROLE_ID_TA_LEVEL_2 (117),
ROLE_ID_TA_LEVEL_15 (126),
ROLE_ID_TA_LEVEL_1 (114)
Drupal Role: TA
Group Role: View

### LINKED ISSUE #20: Make use of form settings values
From: #19 - make modifications needed to ensure private key, public key, and endpoint url are used in building the request.

See also #16

### PR DESCRIPTION
**Pull request by ymdahi**
- [x] Modify form to capture private/public key
- [x] Modify form to capture endpoint url
- [x] #20 
- [x] #16 
- [x] #17 
- [x] #18

### CODE REVIEW

**Comment by ymdahi** [comment]
Next step: modify GroupLMSUserSyncAPI.php to make use of new settings form values.

File: src/Form/GroupLMSUserSyncSettingsForm.php
Comment (danrod96) [inline src/Form/GroupLMSUserSyncSettingsForm.php:1]: I am ok with the changes, can be merged, there's new support for the public & private keys for the new REST API endpoint.

File: src/GroupLMSUserSyncAPI.php
Comment (danrod96) [inline src/GroupLMSUserSyncAPI.php:1]: I see some code refactoring in the file `src/GroupLMSUserSyncAPI.php,` there's some code refactor for the authentication to the API, but the method syncUsersToGroups() is almost completely refactored and It's a bit hard for me to understand what is being changed, I'll approved the merge when it's ready and review from there, for now, I'll make this file as "Viewed"

**Comment by danrod96** [comment]
I mistakenly marked the PR as "Ready for Reviewed", moved it to "Draft State" again.

**Comment by ymdahi** [comment]
Yes, a significant amount of refactoring. This PR brings alignment to UW's middleware. Instead of just a simple update to the form, it actually closes a lot of other tickets as well.

This is good to be merged once you do a sanity check - but it probably won't make too much sense until we align the submodule as well.

**Comment by danrod96** [comment]
Changes are ok, I'll merge the changes and test the module later today, along with the changes in the test submodule, I'll revisit this module once I have the h5p xapi module running.

**Comment by danrod96** [comment]
Can we enable Github Actions for our repos? Just to add some basic checks before merging (code formatting/syntax), I believe I need permissions for that

**Pull request merged**